### ISSUE TITLE
Rewrite `Construct` handling (middleware) in line with `Call`

### ISSUE THREAD

**Issue by iwahbe** [pulumi/pulumi-go-provider#276]
The interface provided by `provider.Provider.Construct` is extremely hard to work with and test against. The interface was "designed" during hack-week and was thus extremely rushed.

https://github.com/pulumi/pulumi-go-provider/blob/c63d6b63a27dfde32bca0f26716207ac195c64ee/provider.go#L1033-L1044

Downsides are:
- Provider authors are never given access to the raw `resource.PropertyMap` inputs and outputs.
- The interface is confusing to use, it's a lambda on a input which must be called to return the output.
- Not compatible with `rpc.Provider` interop: #219
- Not compatible with test mocking: #82

`provider.Provider.Call` has a lot of the same constraints, but a much more reasonable interface:

https://github.com/pulumi/pulumi-go-provider/blob/c63d6b63a27dfde32bca0f26716207ac195c64ee/provider.go#L815-L832

Note that `Call` has the same set of constraints as `Construct`, but there was more time spent doing design work.

To make it easier/possible to support #82, #219 and #266, we should re-write `Construct` in the same style as `Call`.

This will be a breaking change, so it must go in before 1.0.0. This should not affect `infer` based components, since they are authored at a higher level. It will affect how `infer` implements components, and will enable better error messages (or even fixes) to issues like https://github.com/pulumi/pulumi-go-provider/issues/270.

**Comment by EronWright** [comment]
Digging into this, I would guess that one of the challenges during hack-week was related to the `context.Context` parameter that is passed down and eventually baked into the Pulumi context struct.  There's a provider.Wrap function that seeks to replace the context, and the wrapper must be invoked prior to newing up the Pulumi context, else the wrapper's context won't be applied.

**Comment by iwahbe** [comment]
@EronWright Couldn't the wrapper use `(*pulumi.Context).WithValue`?

Commenting on the history, dealing with `context.Context` wasn't the problem during the hack week, when the provider used it's own `infer.Context` struct here.

**Comment by EronWright** [comment]
Good news, I found a reasonable solution (IMO) after a few false starts.  Here's a preview:
https://github.com/pulumi/pulumi-go-provider/compare/main...issue-276-2

The gist of the solution is to move the `Construct` function parameter of `ConstructRequest` into the Go context, to be made available by the provider server to the middleware. This is exposed as a `ProviderHost` interface.

I'm trying to solve for https://github.com/pulumi/pulumi-go-provider/issues/82 and https://github.com/pulumi/pulumi-go-provider/issues/219, and so I imported [some test code](https://github.com/pulumi/pulumi-kubernetes/tree/2d1b080b7df1afb8cfa1cbc92c3bf6576ef5b08e/provider/pkg/host/fake) from pulumi-kubernetes.  With this code, the integration test server (in `integration/`) can mock resource registration to substantively test a component resource.

**Comment by iwahbe** [comment]
The goal of this issue is to rewrite the `Construct` interface to use same abstraction level as the `Call` interface: which is property values (`resource.PropertyMap`). Returning a `pulumi.ComponentResource` does not fit that model.

**Comment by EronWright** [comment]
@iwahbe I agree and the `ConstructRequest` type of `Provider.Construct()` will indeed have property maps just like `Call`.  

Meanwhile, the higher-level `middleware.ComponetResource.Construct()` function should change to be Pulumi Context-oriented.  These two changes simply work well together.

**Comment by iwahbe** [comment]
`middleware.ComponentResource` should mirror the definition of `ConstructRequest`, except at a per-resource level. `infer` should handle the up-leveling part here IMO.
